File: Java/1372. Longest ZigZag Path in a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int ans;

    public int longestZigZag(TreeNode root) {
        ans = 0;
        dfs(root, false, 0);
        dfs(root, true, 0);
        return ans;
    }

    private void dfs(TreeNode node, boolean left, int path) {
        if (node != null) {
            ans = Math.max(ans, path);
            if (left) {
                dfs(node.right, false, path + 1);
                dfs(node.left, true, 1);
            } else {
                dfs(node.left, true, path + 1);
                dfs(node.right, false, 1);
            }
        }
    }
}

File: Python3/1372. Longest ZigZag Path in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        self.path = 0

        def dfs(node, left, path):
            if node:
                self.path = max(self.path, path)

                # current node is left node
                if left:
                    dfs(node.right, False, path + 1)
                    dfs(node.left, True, 1)

                # current node is right node
                else:
                    dfs(node.left, True, path + 1)
                    dfs(node.right, False, 1)

        dfs(root, False, 0)
        dfs(root, True, 0)
        return self.path